File: benchmarks/go/tcp_udp_request_bench.go
// Go TCP & UDP Request Round-Trip Benchmark
// Measures actual request latency: client sends payload, server echoes back

package main

import (
	"fmt"
	"net"
	"time"
)

const N = 1000

func printResults(iterations int, nsElapsed int64, success int) {
	ms := nsElapsed / 1_000_000
	perOp := int64(0)
	opsSec := int64(0)
	if nsElapsed > 0 {
		perOp = nsElapsed / int64(iterations)
		opsSec = int64(iterations) * 1_000_000_000 / nsElapsed
	}
	fmt.Printf("    Iterations: %d\n", iterations)
	fmt.Printf("    Total time: %d ms\n", ms)
	fmt.Printf("    Per op:     %d ns\n", perOp)
	fmt.Printf("    Ops/sec:    %d\n", opsSec)
	fmt.Printf("    Successful: %d/%d\n\n", success, iterations)
}

// ============================================================================
// Benchmark 1: TCP Bind Only (baseline)
// ============================================================================
func benchTcpBind() {
	fmt.Println("=== TCP Bind (baseline) ===")
	fmt.Printf("  %d iterations, bind + close\n\n", N)

	start := time.Now()
	success := 0
	for i := 0; i < N; i++ {
		listener, err := net.Listen("tcp", "127.0.0.1:0")
		if err == nil {
			success++
			listener.Close()
		}
	}
	ns := time.Since(start).Nanoseconds()
	printResults(N, ns, success)
}

// ============================================================================
// Benchmark 2: UDP Bind Only (baseline)
// ============================================================================
func benchUdpBind() {
	fmt.Println("=== UDP Bind (baseline) ===")
	fmt.Printf("  %d iterations, bind + close\n\n", N)

	start := time.Now()
	success := 0
	for i := 0; i < N; i++ {
		conn, err := net.ListenPacket("udp", "127.0.0.1:0")
		if err == nil {
			success++
			conn.Close()
		}
	}
	ns := time.Since(start).Nanoseconds()
	printResults(N, ns, success)
}

// ============================================================================
// Benchmark 3: TCP Request on Reused Connection
// ============================================================================
func benchTcpReusedRequest() {
	fmt.Println("=== TCP Request (reused connection) ===")
	fmt.Printf("  %d iterations, 64-byte payload, echo round-trip\n\n", N)

	// Start echo server
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		fmt.Printf("  ERROR: %v\n\n", err)
		return
	}
	defer listener.Close()
	serverAddr := listener.Addr().String()

	// Echo server goroutine
	go func() {
		conn, err := listener.Accept()
		if err != nil {
			return
		}
		defer conn.Close()
		buf := make([]byte, 256)
		for {
			n, err := conn.Read(buf)
			if err != nil || n == 0 {
				return
			}
			_, err = conn.Write(buf[:n])
			if err != nil {
				return
			}
		}
	}()

	// Give server time to start
	time.Sleep(1 * time.Millisecond)

	// Connect client
	client, err := net.Dial("tcp", serverAddr)
	if err != nil {
		fmt.Printf("  ERROR: %v\n\n", err)
		return
	}
	defer client.Close()

	payload := make([]byte, 64)
	for j := range payload {
		payload[j] = 0x41 // 'A'
	}
	recvBuf := make([]byte, 256)
	success := 0

	start := time.Now()

	for i := 0; i < N; i++ {
		_, err := client.Write(payload)
		if err != nil {
			continue
		}
		n, err := client.Read(recvBuf)
		if err == nil && n > 0 {
			success++
		}
	}

	ns := time.Since(start).Nanoseconds()
	printResults(N, ns, success)
}

// ============================================================================
// Benchmark 4: UDP Request Round-Trip
// ============================================================================
func benchUdpRequest() {
	fmt.Println("=== UDP Request (send + recv echo) ===")
	fmt.Printf("  %d iterations, 64-byte payload, echo round-trip\n\n", N)

	// Bind server
	serverConn, err := net.ListenPacket("udp", "127.0.0.1:0")
	if err != nil {
		fmt.Printf("  ERROR: %v\n\n", err)
		return
	}
	defer serverConn.Close()
	serverAddr := serverConn.LocalAddr()

	// Echo server goroutine
	go func() {
		buf := make([]byte, 256)
		for {
			n, addr, err := serverConn.ReadFrom(buf)
			if err != nil {
				return
			}
			serverConn.WriteTo(buf[:n], addr)
		}
	}()

	// Bind client
	clientConn, err := net.ListenPacket("udp", "127.0.0.1:0")
	if err != nil {
		fmt.Printf("  ERROR: %v\n\n", err)
		return
	}
	defer clientConn.Close()

	payload := make([]byte, 64)
	for j := range payload {
		payload[j] = 0x42 // 'B'
	}
	recvBuf := make([]byte, 256)
	success := 0

	// Give server time to start
	time.Sleep(1 * time.Millisecond)

	start := time.Now()

	for i := 0; i < N; i++ {
		_, err := clientConn.WriteTo(payload, serverAddr)
		if err != nil {
			continue
		}
		n, _, err := clientConn.ReadFrom(recvBuf)
		if err == nil && n > 0 {
			success++
		}
	}

	ns := time.Since(start).Nanoseconds()
	printResults(N, ns, success)
}

// ============================================================================
// Main
// ============================================================================
func main() {
	fmt.Println("\n================================================================")
	fmt.Println("  Go TCP & UDP Request Round-Trip Benchmark")
	fmt.Println("================================================================\n")

	benchTcpBind()
	benchUdpBind()
	benchTcpReusedRequest()
	benchUdpRequest()

	fmt.Println("================================================================")
	fmt.Println("  Notes:")
	fmt.Println("  - TCP reused: single connection, N send+recv round-trips")
	fmt.Println("  - UDP request: send + recv echo round-trip")
	fmt.Println("  - Payload: 64 bytes per request")
	fmt.Println("  - All on 127.0.0.1 (loopback)")
	fmt.Println("================================================================\n")
}

File: benchmarks/rust/tcp_udp_request_bench.rs
// Rust TCP & UDP Request Round-Trip Benchmark
// Measures actual request latency: client sends payload, server echoes back

use std::io::{Read, Write};
use std::net::{TcpListener, TcpStream, UdpSocket};
use std::thread;
use std::time::Instant;

const N: usize = 1000;

fn print_results(iterations: usize, ns_elapsed: i64, success: usize) {
    let ms = ns_elapsed / 1_000_000;
    let per_op = if ns_elapsed > 0 { ns_elapsed / iterations as i64 } else { 0 };
    let ops_sec = if ns_elapsed > 0 {
        (iterations as i128 * 1_000_000_000) / ns_elapsed as i128
    } else {
        0
    };
    println!("    Iterations: {}", iterations);
    println!("    Total time: {} ms", ms);
    println!("    Per op:     {} ns", per_op);
    println!("    Ops/sec:    {}", ops_sec);
    println!("    Successful: {}/{}\n", success, iterations);
}

// ============================================================================
// Benchmark 1: TCP Bind Only (baseline)
// ============================================================================
fn bench_tcp_bind() {
    println!("=== TCP Bind (baseline) ===");
    println!("  {} iterations, bind + close\n", N);

    let start = Instant::now();
    let mut success = 0;
    for _ in 0..N {
        if let Ok(_listener) = TcpListener::bind("127.0.0.1:0") {
            success += 1;
        }
    }
    let ns = start.elapsed().as_nanos() as i64;
    print_results(N, ns, success);
}

// ============================================================================
// Benchmark 2: UDP Bind Only (baseline)
// ============================================================================
fn bench_udp_bind() {
    println!("=== UDP Bind (baseline) ===");
    println!("  {} iterations, bind + close\n", N);

    let start = Instant::now();
    let mut success = 0;
    for _ in 0..N {
        if let Ok(_socket) = UdpSocket::bind("127.0.0.1:0") {
            success += 1;
        }
    }
    let ns = start.elapsed().as_nanos() as i64;
    print_results(N, ns, success);
}

// ============================================================================
// Benchmark 3: TCP Request on Reused Connection
// ============================================================================
fn bench_tcp_reused_request() {
    println!("=== TCP Request (reused connection) ===");
    println!("  {} iterations, 64-byte payload, echo round-trip\n", N);

    let listener = match TcpListener::bind("127.0.0.1:0") {
        Ok(l) => l,
        Err(e) => {
            println!("  ERROR: {}\n", e);
            return;
        }
    };
    let server_addr = listener.local_addr().unwrap();

    // Echo server thread
    thread::spawn(move || {
        if let Ok((mut stream, _)) = listener.accept() {
            let mut buf = [0u8; 256];
            loop {
                match stream.read(&mut buf) {
                    Ok(0) => break,
                    Ok(n) => {
                        if stream.write_all(&buf[..n]).is_err() {
                            break;
                        }
                    }
                    Err(_) => break,
                }
            }
        }
    });

    // Small delay for server to start
    thread::sleep(std::time::Duration::from_millis(1));

    let mut client = match TcpStream::connect(server_addr) {
        Ok(s) => s,
        Err(e) => {
            println!("  ERROR: {}\n", e);
            return;
        }
    };

    let payload = [0x41u8; 64]; // 64 bytes of 'A'
    let mut recv_buf = [0u8; 256];
    let mut success = 0;

    let start = Instant::now();

    for _ in 0..N {
        if client.write_all(&payload).is_ok() {
            if let Ok(n) = client.read(&mut recv_buf) {
                if n > 0 {
                    success += 1;
                }
            }
        }
    }

    let ns = start.elapsed().as_nanos() as i64;
    print_results(N, ns, success);
}

// ============================================================================
// Benchmark 4: UDP Request Round-Trip
// ============================================================================
fn bench_udp_request() {
    println!("=== UDP Request (send + recv echo) ===");
    println!("  {} iterations, 64-byte payload, echo round-trip\n", N);

    let server = match UdpSocket::bind("127.0.0.1:0") {
        Ok(s) => s,
        Err(e) => {
            println!("  ERROR: {}\n", e);
            return;
        }
    };
    let server_addr = server.local_addr().unwrap();

    // Echo server thread
    let server_clone = server.try_clone().unwrap();
    thread::spawn(move || {
        let mut buf = [0u8; 256];
        loop {
            match server_clone.recv_from(&mut buf) {
                Ok((n, addr)) => {
                    let _ = server_clone.send_to(&buf[..n], addr);
                }
                Err(_) => break,
            }
        }
    });

    let client = match UdpSocket::bind("127.0.0.1:0") {
        Ok(s) => s,
        Err(e) => {
            println!("  ERROR: {}\n", e);
            return;
        }
    };

    let payload = [0x42u8; 64]; // 64 bytes of 'B'
    let mut recv_buf = [0u8; 256];
    let mut success = 0;

    // Small delay for server to start
    thread::sleep(std::time::Duration::from_millis(1));

    let start = Instant::now();

    for _ in 0..N {
        if client.send_to(&payload, server_addr).is_ok() {
            if let Ok((n, _)) = client.recv_from(&mut recv_buf) {
                if n > 0 {
                    success += 1;
                }
            }
        }
    }

    let ns = start.elapsed().as_nanos() as i64;
    print_results(N, ns, success);
}

// ============================================================================
// Main
// ============================================================================
fn main() {
    println!("\n================================================================");
    println!("  Rust TCP & UDP Request Round-Trip Benchmark");
    println!("================================================================\n");

    bench_tcp_bind();
    bench_udp_bind();
    bench_tcp_reused_request();
    bench_udp_request();

    println!("================================================================");
    println!("  Notes:");
    println!("  - TCP reused: single connection, N send+recv round-trips");
    println!("  - UDP request: send + recv echo round-trip");
    println!("  - Payload: 64 bytes per request");
    println!("  - All on 127.0.0.1 (loopback)");
    println!("================================================================\n");
}
